### PR TITLE
feat(workbench): origin profiles — full CRUD, Rust parity, and integration fixes

### PR DESCRIPTION
**Pull request by bb-connor**
## Summary
- **Origins page rewrite**: Fully editable origin profiles with localStorage-backed CRUD, search/filter, clone, reorder (priority-based matching), and delete with referential integrity checks against saved policies
- **Rust engine parity**: require_confirmation for MCP tools, merge-mode fields (additional_allow/block, remove_allow/block), actor_id/actor_type in match rules, metadata JSON editor, max_args_size, custom provider/space_type support, bridge targets editor
- **Integration fixes**: Profile ID rename cascading to saved/active policies, type alignment between approval-types and canonical OriginContext, provider coverage (email/discord/webhook) in fleet-client and approval-queue, Visibility type expansion, hierarchy-page union narrowing fix
- **Code quality**: Rules of Hooks fix in origin-editor, stable React keys for bridge targets and profile list items, case-insensitive ID validation, search always shows selected profile

## Key files changed
| File | Changes |
|------|---------|
| `origins/origins-page.tsx` | New — full profile library with inline editing |
| `editor/origin-editor.tsx` | MCP merge fields, custom providers, metadata, actor fields |
| `lib/workbench/types.ts` | McpToolConfig, OriginMatch, OriginProfile, Visibility expanded |
| `lib/workbench/approval-types.ts` | Extends canonical OriginContext instead of hardcoded types |
| `lib/workbench/fleet-client.ts` | Provider coverage + proper field extraction |
| `approvals/approval-queue.tsx` | Email/discord/webhook provider UI config |
| `hierarchy/hierarchy-page.tsx` | Type error fix (union narrowing) |

## Test plan
- [ ] Navigate to /origins — profiles should load from localStorage, be fully editable
- [ ] Create, clone, rename, reorder, and delete profiles; verify persistence across page reloads
- [ ] Apply a profile to active policy, rename it, confirm cascade updates policy
- [ ] Delete a profile referenced by a policy — confirm warning dialog appears
- [ ] Toggle MCP/Egress sections on/off, verify enabled flag and data preservation
- [ ] Enter custom provider/space_type via "Custom..." option
- [ ] Edit metadata JSON field — verify validation on blur
- [ ] Navigate to /editor — origin editor should show matching parity fields
- [ ] Verify /approvals page renders email/discord/webhook provider badges
- [ ] Type-check passes: `npx tsc --noEmit` (excluding pre-existing test matcher issues)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches core hierarchy push/pull/sync flows (including optimistic rollback and ID remapping), so regressions could desync local state from fleet data. UI/editor changes are straightforward but broaden the configuration surface area and require careful validation of edge cases.
> 
> **Overview**
> **Origin profiles UI gets expanded editing and better state handling.** `OriginEditor` now sources option lists from `origin-options`, fixes a Rules-of-Hooks issue, tolerates missing `profiles`, adds custom provider/space-type modes, adds match fields (`thread_id`, `actor_id`, `actor_type`), and introduces a validated JSON `metadata` editor; approvals UI also recognizes new origin providers (email/discord/webhook).
> 
> **Policy overrides surface more Rust-parity knobs.** MCP overrides add `require_confirmation`, `max_args_size`, and merge-style allow/block deltas; egress overrides add `log` as a default action.
> 
> **Hierarchy page shifts to endpoint/runtime leaves and hardens fleet sync.** The tree supports `endpoint`/`runtime` node types, updates leaf counting/labels, disables drag/actions during sync, normalizes loaded/pulled hierarchies to preserve child order, preserves/pushes `external_id`, and adds sync guards: confirmation prompts, in-progress locking, optimistic rollback with versioning to avoid stale reverts, and remapping local IDs to server-assigned IDs on create/push.
> 
> **Tests added/expanded** for origin editor state synchronization/custom modes and for hierarchy normalization, legacy node handling, push/pull edge cases, ID remapping, and concurrent rollback ordering.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4b19f58b414b1932886c12c269f179daf4145a46. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by bb-connor** [comment]
@codex

File: apps/workbench/src/lib/workbench/multi-policy-store.tsx [hunk 1]
```diff
@@ -839,6 +840,10 @@ interface PersistedTabState {
   activeTabId: string;
 }
 
+/** Fields that must never be persisted to localStorage. */
+const SENSITIVE_GUARD_FIELDS = new Set(["embedding_api_key"]);
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply apps/workbench/src/lib/workbench/multi-policy-store.tsx]: Removed in `11aa0b314`. The store already relies on `sanitizeYamlForStorageWithMetadata`, so `SENSITIVE_GUARD_FIELDS` was dead code and has been deleted.

File: apps/workbench/src/components/workbench/hierarchy/hierarchy-page.tsx [hunk 2]
```diff
@@ -244,13 +257,25 @@ function TreeNode({
...
+            )}
+            {canAddRuntime && (
               <button
                 className="p-0.5 rounded hover:bg-[#3dbf84]/20 text-[#3dbf84]/60 hover:text-[#3dbf84]"
-                title="Add agent"
+                title="Add Runtime"
                 onClick={(e) => {
                   e.stopPropagation();
-                  onAddChild(node.id, "agent");
+                  onAddChild(node.id, "runtime");
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply apps/workbench/src/components/workbench/hierarchy/hierarchy-page.tsx]: Handled in `11aa0b314`. `getLeafAgents` now treats `agent`, `runtime`, and endpoint-without-runtime nodes as leaf enforcement targets, and the hierarchy page now renders generic leaf counts/affected leaves instead of runtime-only labels. I also added hierarchy regressions for legacy agent leaves, runtime leaves, and `external_id` round-tripping in `hierarchy-page.test.tsx`.

File: apps/workbench/src/components/workbench/origins/origins-page.tsx [hunk 3]
```diff
@@ -0,0 +1,2647 @@
...
+        for (const sp of state.savedPolicies) {
+          const profiles = sp.policy.origins?.profiles;
+          if (profiles?.some((p) => p.id === oldId)) {
+            const updatedProfiles = profiles.map((p) =>
+              p.id === oldId ? { ...p, id: newId } : p,
+            );
+            const updatedSavedPolicy = {
+              ...sp,
+              policy: {
+                ...sp.policy,
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply apps/workbench/src/components/workbench/origins/origins-page.tsx]: Fixed in `11aa0b314` by moving the rename cascade through `origin-profile-utils.ts`. `renameOriginProfileIdInSavedPolicy(...)` now rebuilds `savedPolicy.yaml` from the updated policy before `SAVE_POLICY`, and `origin-profile-utils.test.ts` covers that YAML regeneration path.

File: apps/workbench/src/components/workbench/hierarchy/hierarchy-page.tsx [hunk 4]
```diff
@@ -1395,20 +1557,22 @@ export function HierarchyPage() {
...
    */
   const mapNodeType = useCallback((backendType: string): OrgNodeType => {
-    if (backendType === "org" || backendType === "team" || backendType === "agent") {
+    if (backendType === "org" || backendType === "team" || backendType === "endpoint" || backendType === "runtime") {
       return backendType;
     }
-    // "project" and any unknown types map to "team"
-    return "team";
+    if (backendType === "agent") return "endpoint"; // backward compat
+    return "team"; // "project" and unknown
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply apps/workbench/src/components/workbench/hierarchy/hierarchy-page.tsx]: Fixed in `11aa0b314`. Live hierarchy pull now preserves backend `agent` nodes as frontend `agent` nodes instead of coercing them to `endpoint`, so legacy leaves still participate in impact preview and validation. I added a regression in `hierarchy-page.test.tsx` for the live-pull legacy-agent case.

**Comment by bb-connor** [comment]
Follow-up in `d7354b97f` addresses the remaining workbench review gaps: hierarchy push now distinguishes success-without-id from hard failures, legacy `agent -> team` drag/drop is restored, origin profile list items use stable keys, both origin editors expose the full `Visibility` enum, the editor metadata textarea resyncs from upstream profile changes, and persisted browser state no longer stores `testSuiteYaml`.

**Comment by bb-connor** [comment]
@codex

**Comment by bb-connor** [comment]
@codex

**Comment by bb-connor** [comment]
@codex
